cpp/memilio/io/history.h

```diff
@@ -43,21 +43,31 @@ constexpr size_t index_templ_pack()
 }
 } // namespace details
 
-struct LogOnce {
+/* These are the base classes for Loggers. They are used to determine when a logger is called. 
+* LogOnceStart is only passed to Writer on the first call to History::log, LogAlways on all calls.
+* LogIf is only logging to Writer if the should_log is true. This needs to be implemented, e.g. for logging only the last step.
+*/
+struct LogOnceStart {
 };
 
 struct LogAlways {
 };
 
+template <typename Cond>
+struct LogIf {
```

Comment:
Thats a good point. The reason we did not remove LogOnce is that it needs a cached value, i.e. a bool indicating whether 'once' already happened. This is stored in the History object, since we wanted to have static loggers, so that we only need to initialize the data, and not worry about managing logger objects.

I think we can probably drop the static requirement, and instead only require that a class logger has a log and should_log function, and is default constructible/destructible. Then the history can keep an additional tuple of loggers, and just call the member functions (i.e. `logger_instance.log()` instead of `Logger::log()` ).
This also allows the loggers to allocate (static) variables, and removes the requirement to inherit from one of the Log* structs (but we can still provide them as examples/defaults like you mentioned).

We could alternatively implement a caching mechanism, but I think non-static loggers is a more future-proof approach.

Reply:
We can also add a function `History::log<Logger>(const T& t)` to allow manually using a single logger. It is trivial to implement, but do we actually need or want it?